### PR TITLE
Drush command to create Apigee Edge role

### LINKED ISSUE #210: Add Edge role creation script to module
Currently the [module's documentation installation steps](https://www.drupal.org/docs/8/modules/apigee-edge/configure-the-connection-to-apigee-edge#assigning-the-developer-administrator-role) say to download a gist from my GitHub repo, which is weird.  This should be instead a script in the module.

### PR DESCRIPTION
**Pull request by cnovak**
Fixes #210

apigee-edge:create-edge-role Drush command will create role will correct permissions for Apigee Edge connection.

### CODE REVIEW

**Comment by cnovak** [comment]
Tests for this module are failing: https://travis-ci.org/cnovak/apigee-edge-drupal/builds/547549135

Here is screenshot of all tests added running properly locally:

Screen Shot 2019-06-19 at 4 41 27 PM

File: src/Commands/ApigeeEdgeCommands.php [hunk 1]
```diff
@@ -19,29 +19,43 @@
...
+use Drupal\apigee_edge\Util\EdgeConnectionUtilServiceInterface;
 use Drush\Commands\DrushCommands;
 
 /**
  * Drush 9 command file.
  */
 class ApigeeEdgeCommands extends DrushCommands {
 
+  // Default base url.
+  const DEFAULT_BASE_URL = 'https://api.enterprise.apigee.com/v1';
```
Comment (mxr576) [inline src/Commands/ApigeeEdgeCommands.php]: This information belongs to the API client that takes care of the role creation, not to this base class. (See my comment below about re-using the API client from the PHP API client.)

File: src/Util/EdgeConnectionUtilService.php [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  protected $httpClient;
+
+  /**
+   * Constructs a EdgeRoleCreator object.
+   *
+   * @param \GuzzleHttp\ClientInterface $http_client
+   *   The HTTP client.
+   */
+  public function __construct(ClientInterface $http_client) {
+    $this->httpClient = $http_client;
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilService.php]: Even if the PHP API client does not support this API, you can still use it to communicate with this MGMT API. You do not need to start from scratch. The PHP API Client's API client can also provide you better HTTP error handling out of the box.

File: src/Util/EdgeConnectionUtilService.php [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge\Util;
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilService.php]: Move this to the `Drupal\apigee_edge\Command\Util\` namespace or something similar otherwise external developers may get tempted to use this in Drupal modules instead of the SDKConnector service.

File: src/Util/EdgeConnectionUtilService.php [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+  /**
+   * The HTTP client.
+   *
+   * @var \GuzzleHttp\Client
+   */
+  protected $httpClient;
+
+  /**
+   * Constructs a EdgeRoleCreator object.
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilService.php]: Outdated docblock.

File: src/Util/EdgeConnectionUtilServiceInterface.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+namespace Drupal\apigee_edge\Util;
+
+use Symfony\Component\Console\Style\StyleInterface;
+
+/**
+ * Defines an interface for Edge connection classes.
+ */
+interface EdgeConnectionUtilServiceInterface {
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilServiceInterface.php]: We probably do not want anybody to replace the implementation shipped by this module so we do not need an interface for this.

File: src/Util/EdgeConnectionUtilService.php [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        return;
+      }
+    }
+    else {
+      $io->text('Role ' . $role_name . ' already exists.');
+    }
+
+    $this->setDefaultPermissions($io, $t, $org, $email, $password, $base_url, $role_name);
+
+    $io->success($t('Role !role is configured.  Log into apigee.com to assign a user to this role.', ['!role' => $role_name]));
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilService.php]: Unnecessary extra space: `ed.  Log i`

File: src/Commands/ApigeeEdgeCommands.php [hunk 7]
```diff
@@ -19,29 +19,43 @@
...
+
+  // Default role name to create in Apigee Edge.
+  const DEFAULT_ROLE_NAME = 'drupalportal';
```
Comment (mxr576) [inline src/Commands/ApigeeEdgeCommands.php]: This information belongs to the service that takes care of the role creation, not to this base class.

File: src/Commands/ApigeeEdgeCommands.php [hunk 8]
```diff
@@ -57,4 +71,61 @@ public function sync() {
...
+   */
+  public function createEdgeRole(
+    string $org,
+    string $email,
+    array $options = [
+      'password' => NULL,
+      'base-url' => self::DEFAULT_BASE_URL,
+      'role-name' => self::DEFAULT_ROLE_NAME,
+    ]) {
+    $this->edgeConnectionUtilService->createEdgeRoleForDrupal($this->io(), 'dt', $org, $email, $options['password'], $options['base-url'], $options['role-name']);
```
Comment (mxr576) [inline src/Commands/ApigeeEdgeCommands.php]: User input should be validated, ex.: base-url should be a valid URI, role name should be string, etc. Symfony's options resolver could handle this easily.

File: src/Util/EdgeConnectionUtilService.php [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Exception\TransferException;
+use Symfony\Component\Console\Style\StyleInterface;
+
+/**
+ * Defines an interface for Edge connection classes.
+ */
+class EdgeConnectionUtilService implements EdgeConnectionUtilServiceInterface {
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilService.php]: How about calling this as `ApigeeEdgeManagementCliService` or something else that provides a more descriptive and also generic name? I add `Cli` because this service (class) should be used only by CLIs tools.

File: src/Util/EdgeConnectionUtilService.php [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+      '/apiproducts' => ['get', 'put'],
+      // GET, PUT, & DELETE access.
+      '/developers' => ['get', 'put', 'delete'],
+      '/developers/*/apps' => ['get', 'put', 'delete'],
+      '/developers/*/apps/*' => ['get', 'put', 'delete'],
+      '/companies/*' => ['get', 'put', 'delete'],
+      '/companies/*/apps/*' => ['get', 'put', 'delete'],
+    ];
+
+    $url = $base_url . '/o/' . $org . '/userroles/' . $role_name . '/permissions';
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilService.php]: This form is better and easier readable `"/o/{$org}/userroles/{$role_name}/permissions";`

File: src/Util/EdgeConnectionUtilService.php [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+   * @param string $base_url
+   *   The base url of the Edge API.
+   * @param string $role_name
+   *   The role name to add the permissions to.
+   */
+  protected function setDefaultPermissions(StyleInterface $io, callable $t, string $org, string $email, string $password, string $base_url, string $role_name) {
+    $io->text('Setting permissions on role ' . $role_name . '.');
+
+    $permissions = [
+      // GET access.
```
Comment (mxr576) [inline src/Util/EdgeConnectionUtilService.php]: Here you should rather describe in comments why we grant only `GET` or only `GET & PUT` or `GET, PUT & DELETE` access to certain API endpoints. These comments could provide more value than the current ones for later when someone needs to change something on this permission matrix.

**Review by mxr576 — CHANGES_REQUESTED** [review]
Did not do a complete review, I've just I shared some initial suggestions.
Just like the sync supports both Drupal Console and Drush, this new command should also support Drupal Console as well. See `console.services.yml`

File: apigee_edge.services.yml [hunk 12]
```diff
@@ -13,6 +13,10 @@ services:
   apigee_edge.cli:
     class: Drupal\apigee_edge\CliService
 
+  apigee_edge.edge_connection_util:
```
Comment (mxr576) [inline apigee_edge.services.yml]: IIRC we planned to use the `apigee_edge.cli` service as the source for all CLI commands that this module provides. Maybe you should add the create role command to this service and that way we do not need to worry about how we should call this new service.

File: src/Util/EdgeConnectionUtilService.php (hunk 3 above)
Comment (cnovak) [reply src/Util/EdgeConnectionUtilService.php]: The Drush commands will need to also use this directory, but that works for me

File: src/Util/EdgeConnectionUtilService.php (hunk 4 above)
Comment (cnovak) [reply src/Util/EdgeConnectionUtilService.php]: done

File: src/Util/EdgeConnectionUtilServiceInterface.php (hunk 5 above)
Comment (cnovak) [reply src/Util/EdgeConnectionUtilServiceInterface.php]: Always good for testing purposes to have interfaces

File: src/Util/EdgeConnectionUtilService.php (hunk 6 above)
Comment (cnovak) [reply src/Util/EdgeConnectionUtilService.php]: done

File: src/Commands/ApigeeEdgeCommands.php (hunk 1 above)
Comment (cnovak) [reply src/Commands/ApigeeEdgeCommands.php]: Moved to `EdgeConnectionUtilServiceInterface`

File: src/Commands/ApigeeEdgeCommands.php (hunk 7 above)
Comment (cnovak) [reply src/Commands/ApigeeEdgeCommands.php]: Moved to `EdgeConnectionUtilServiceInterface`

File: src/Commands/ApigeeEdgeCommands.php (hunk 8 above)
Comment (cnovak) [reply src/Commands/ApigeeEdgeCommands.php]: Turns out that role and other params can be a number and not sure what the rules are, so we can let the Apigee API respond back with any errors. Added validation for Base URL however.

File: src/Util/EdgeConnectionUtilService.php (hunk 10 above)
Comment (cnovak) [reply src/Util/EdgeConnectionUtilService.php]: Done

File: src/Util/EdgeConnectionUtilService.php (hunk 11 above)
Comment (cnovak) [reply src/Util/EdgeConnectionUtilService.php]: Done

File: apigee_edge.services.yml (hunk 12 above)
Comment (cnovak) [reply apigee_edge.services.yml]: Done

File: src/Util/EdgeConnectionUtilService.php (hunk 9 above)
Comment (cnovak) [reply src/Util/EdgeConnectionUtilService.php]: good idea, done

File: src/Command/CreateEdgeRoleCommand.php [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        'p',
+        InputArgument::OPTIONAL,
+        $this->trans('commands.apigee_edge.role.create.options.password')
+      )
+      ->addOption(
+      'base-url',
+      'b',
+      InputArgument::OPTIONAL,
+      $this->trans('commands.apigee_edge.role.create.options.base-url')
+    )->addOption(
```
Comment (kscheirer) [inline src/Command/CreateEdgeRoleCommand.php]: needs more indent

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+   */
+  public function createEdgeRoleForDrupal(StyleInterface $io,
+                                          callable $t,
+                                          string $org,
+                                          string $email,
+                                          string $password,
+                                          ?string $base_url,
+                                          ?string $role_name) {
+
+    if ($base_url === NULL || $base_url === '') {
```
Comment (kscheirer) [inline src/Command/Util/ApigeeEdgeManagementCliService.php]: `empty($base_url)` is a little shorter, unless you really only want to test for those 2 conditions.

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+      // Set default base URL if var is null or empty string.
+      $base_url = self::DEFAULT_BASE_URL;
+    }
+    else {
+      // Validate it is a valid URL.
+      if (!UrlHelper::isValid($base_url, TRUE)) {
+        $io->error($t('Base URL is not valid.'));
+        return;
+      }
+
```
Comment (kscheirer) [inline src/Command/Util/ApigeeEdgeManagementCliService.php]: extra blank line

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+    if (!$does_role_exist) {
+      $io->text($t('Role :role does not exist. Creating role.', [':role' => $role_name]));
+
+      $url = "{$base_url}/o/{$org}/userroles";
+      try {
+        $this->httpClient->post($url, [
+          'body' => json_encode([
+            'role' => [$role_name],
+          ]),
+          'auth' => [$email, $password],
```
Comment (kscheirer) [inline src/Command/Util/ApigeeEdgeManagementCliService.php:101]: This won't support oauth-enabled orgs I think. Is there a way to use the php-sdk instead here, which should handle that for you?

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+      catch (TransferException $exception) {
+        $this->handleHttpClientExceptions($exception, $io, $t, $url, $org, $email);
+        return;
+      }
+    }
+    else {
+      $io->text('Role ' . $role_name . ' already exists.');
+    }
+
+    $this->setDefaultPermissions($io, $t, $org, $email, $password, $base_url, $role_name);
```
Comment (kscheirer) [inline src/Command/Util/ApigeeEdgeManagementCliService.php:114]: This could potentially reset some permissions if the role already exists and had its permissions modified from default. Maybe only set defaults if the role was just created? ie move into `try {}` block

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+        'auth' => [$email, $password],
+        'headers' => ['Accept' => 'application/json'],
+      ]);
+    }
+    catch (TransferException $exception) {
+      $this->handleHttpClientExceptions($exception, $io, $t, $url, $org, $email);
+      return FALSE;
+    }
+
+    $body = json_decode($response->getBody());
```
Comment (kscheirer) [inline src/Command/Util/ApigeeEdgeManagementCliService.php]: `JSON_THROW_ON_ERROR` would be nice here, but requires php 7.3 :(

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+   * @param string $url
+   *   The url being connected to.
+   * @param string $org
+   *   The organization to connect to.
+   * @param string $email
+   *   The email of an Edge user with org admin role to make Edge API calls.
+   */
+  public function handleHttpClientExceptions(TransferException $exception, StyleInterface $io, callable $t, string $url, string $org, string $email): void {
+    $io->error($t('Error connecting to Apigee Edge. :exception_message', [':exception_message' => $exception->getMessage()]));
+    switch ($exception->getCode()) {
```
Comment (kscheirer) [inline src/Command/Util/ApigeeEdgeManagementCliService.php:316]: add a default case to handle when the error is not 0,401,403,302?

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 20]
```diff
@@ -0,0 +1,330 @@
...
+        ]));
+        return;
+
+      case 302:
+        $io->note($t('Edge endpoint gives a redirect response, is the url :url does not seem to be a valid Apigee Edge endpoint.', [':url' => $url]));
+        return;
+
+    }
+  }
+
```
Comment (kscheirer) [inline src/Command/Util/ApigeeEdgeManagementCliService.php:339]: extra blank line

**Review by kscheirer — COMMENTED** [review]
Looks pretty good, only supporting basic auth seems like an issue though. Better than supporting no auths though :)

File: src/Command/CreateEdgeRoleCommand.php [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * @Drupal\Console\Annotations\DrupalCommand (
+ *     extension="apigee_edge",
+ *     extensionType="module"
+ * )
+ */
+class CreateEdgeRoleCommand extends CommandBase {
+
+  // Default base url.
+  const DEFAULT_BASE_URL = 'https://api.enterprise.apigee.com/v1';
```
Comment (arlina-espinoza) [inline src/Command/CreateEdgeRoleCommand.php]: It would be better to use the default endpoint from the PHP client library (`Apigee\Edge\ClientInterface::DEFAULT_ENDPOINT`) instead of duplicating the value here.

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 22]
```diff
@@ -0,0 +1,330 @@
...
+        $io->error($t('Base URL is not valid.'));
+        return;
+      }
+
+    }
+
+    if ($role_name === NULL || $role_name === '') {
+      // Set default if null or empty string.
+      $role_name = self::DEFAULT_ROLE_NAME;
+    }
```
Comment (arlina-espinoza) [inline src/Command/Util/ApigeeEdgeManagementCliService.php]: Just a suggestion: `$role_name = $role_name ?: self::DEFAULT_ROLE_NAME;`

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 17 above)
Comment (arlina-espinoza) [reply src/Command/Util/ApigeeEdgeManagementCliService.php:114]: Maybe adding another command line option to "force" resetting permissions if the role has already been created?

File: src/Command/Util/ApigeeEdgeManagementCliService.php [hunk 23]
```diff
@@ -0,0 +1,330 @@
...
+      '/' => ['get'],
+      // Read only access to environments.
+      '/environments/' => ['get'],
+      '/environments/*/stats/*' => ['get'],
+      // We do not need to update/edit roles, just read them.
+      '/userroles' => ['get'],
+      // No need to create API products, only read and edit.
+      '/apiproducts' => ['get', 'put'],
+      // Full CRUD for developers.
+      '/developers' => ['get', 'put', 'delete'],
```
Comment (arlina-espinoza) [inline src/Command/Util/ApigeeEdgeManagementCliService.php:151]: I think this is missing the `/developers/*` path, it would be:
```
'/developers' => ['get', 'put'],
'/developers/*' => ['get', 'put', 'delete'],
```

Comment (arlina-espinoza) [reply src/Command/Util/ApigeeEdgeManagementCliService.php:151]: Sorry, further reading of the docs (https://docs.apigee.com/api-platform/system-administration/managing-roles-api#aboutpermissionsetting-syntaxandprecedence) made me see your code is correct - disregard previous comment.

File: src/Command/Util/ApigeeEdgeManagementCliServiceInterface.php [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+
+use Symfony\Component\Console\Style\StyleInterface;
+
+/**
+ * Defines an interface for Edge connection classes.
+ */
+interface ApigeeEdgeManagementCliServiceInterface {
+
+  // Default base url.
+  const DEFAULT_BASE_URL = 'https://api.enterprise.apigee.com/v1';
```
Comment (arlina-espinoza) [inline src/Command/Util/ApigeeEdgeManagementCliServiceInterface.php]: It would be better to use the constant from `Apigee\Edge\ClientInterface::DEFAULT_ENDPOINT`.

File: tests/src/Functional/ApiProductAccessTest.php [hunk 25]
```diff
@@ -306,8 +306,8 @@ protected function developerAppEditFormTest() {
 
     // >> Bypass user.
     $this->drupalLogin($this->users[self::USER_WITH_BYPASS_PERM]);
-    // Even if a user has bypass permission it should see only those API
-    // Products on an other user's add/edit form that the other user has
+    // Even if a user has bypass permission they should see only those API
```
Comment (arlina-espinoza) [inline tests/src/Functional/ApiProductAccessTest.php:309]: Changes in this file (tests/src/Functional/ApiProductAccessTest.php) seem unrelated to this PR - although I'm ok with adding them as they are only comment typo fixes.

File: tests/src/Unit/Command/CreateEdgeRoleCommandTest.php [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function setUp() {
+      parent::setUp();
+      $this->cliService = $this->prophesize(CliServiceInterface::class);
+      $this->logMessageParser = $this->prophesize(LogMessageParserInterface::class);
+      $this->loggerChannelFactory = $this->prophesize(LoggerChannelFactoryInterface::class);
+      $this->createEdgeRoleCommand = new CreateEdgeRoleCommand($this->cliService->reveal(),
```
Comment (arlina-espinoza) [inline tests/src/Unit/Command/CreateEdgeRoleCommandTest.php:104]: Running this test locally gives me an error because Edge does not require Drupal Console. Trying it out in travis right now, but I think we should at least add Drupal Console to `.travis/composer.json`.

**Review by arlina-espinoza — CHANGES_REQUESTED** [review]
I've added some comments. I think besides @kscheirer suggestions, we should add Drupal Console to travis' composer build, or to the module's require-dev. We had talked before about adding it to `apigee/devportal-kickstart-project-composer`, but I think it is needed here as well.

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 17 above)
Comment (arlina-espinoza) [reply src/Command/Util/ApigeeEdgeManagementCliService.php:114]: Also per the docs, this call "does not remove any existing permissions in the role", but I agree with @kscheirer that calling it for an existing role (which might have been altered outside of this call) might create the  impression that this command validated the role to only have the required permissions. I think that a) either exit with error if the role already exists, or b) force resetting the role, which would involve deleting all permissions and adding the right ones again.

File: src/Util/EdgeConnectionUtilService.php (hunk 2 above)
Comment (arlina-espinoza) [reply src/Util/EdgeConnectionUtilService.php]: This would be an interesting addition. I think using the PHP API client would be more consistent. The `apigee_edge.sdk_connector` service has a `getClient()` method that you can pass in a custom auth and endpoint, so maybe that can work.
Also, I think adding these API endpoints to the management API would be better in the long run - although not a blocker for adding this here.

Comment (cnovak) [reply src/Util/EdgeConnectionUtilService.php]: I added a new issue for this #274 so that we can handle it in a new PR.

**Review by cnovak — COMMENTED** [review]
> Looks pretty good, only supporting basic auth seems like an issue though. Better than supporting no auths though :)

Agreed, new issue #274 will handle this.

File: src/Command/CreateEdgeRoleCommand.php (hunk 13 above)
Comment (cnovak) [reply src/Command/CreateEdgeRoleCommand.php]: done

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 14 above)
Comment (cnovak) [reply src/Command/Util/ApigeeEdgeManagementCliService.php]: done

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 15 above)
Comment (cnovak) [reply src/Command/Util/ApigeeEdgeManagementCliService.php]: Are you not suppose to have blank lines in if/else? I changed it however.

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 16 above)
Comment (cnovak) [reply src/Command/Util/ApigeeEdgeManagementCliService.php:101]: new issue #274 will handle this.

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 17 above)
Comment (cnovak) [reply src/Command/Util/ApigeeEdgeManagementCliService.php:114]: Was thinking this was OK, but you both bring up points that someone may make changes to the role so we do not want to stomp those changes.

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 18 above)
Comment (cnovak) [reply src/Command/Util/ApigeeEdgeManagementCliService.php]: Agree, but we would need to up the PHP version allowed for our module, I don't think we should restrict it that much yet

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 19 above)
Comment (cnovak) [reply src/Command/Util/ApigeeEdgeManagementCliService.php:316]: Line 304 `$io->error()` will display out info about any error, the switch statement is only to add extra info for the end user for common use cases. Here is example of output from response status code of 506:

Screen Shot 2019-10-04 at 3 53 57 PM

Adding comment to section to clarify.

File: tests/src/Functional/ApiProductAccessTest.php (hunk 25 above)
Comment (cnovak) [reply tests/src/Functional/ApiProductAccessTest.php:309]: This is unrelated, did a comment typo fix.

File: tests/src/Unit/Command/CreateEdgeRoleCommandTest.php (hunk 26 above)
Comment (cnovak) [reply tests/src/Unit/Command/CreateEdgeRoleCommandTest.php:104]: I will also add drush which is missing

File: src/Command/Util/ApigeeEdgeManagementCliService.php (hunk 17 above)
Comment (cnovak) [reply src/Command/Util/ApigeeEdgeManagementCliService.php:114]: Added a `--force` flag

**Comment by cnovak** [comment]
Added changes, this PR needs to be looked over again to see if it should be approved

**Comment by cnovak** [comment]
@arlina-espinoza the merge conflict is now fixed, ready for review again.

File: console/translations/en/apigee_edge.role.create.yml [hunk 27]
```diff
@@ -0,0 +1,13 @@
+description: 'Create Apigee role for Drupal'
+help: 'Create a custom Apigee role that limits permissions for Drupal connections to the Apigee API.'
+arguments:
+  org: 'The Apigee Edge org to create the role in'
+  email: 'An Apigee user email address with orgadmin role for this org'
+options:
+  password: 'Password for the Apigee orgadmin user (if not passed in you will be prompted)'
+  base-url: 'Base URL to use, defaults to public cloud URL https://api.enterprise.apigee.com/v1'
+  role-name: 'The role to create in the Apigee Edge org, defaults to "drupalportal"'
+  force: 'Force running of permissions on a role that already exists.'
```
Comment (arlina-espinoza) [inline console/translations/en/apigee_edge.role.create.yml]: Does this mean that it will add the permissions, or reset the permissions of the role if it already exists? The wording does not make that clear.

File: tests/src/Functional/ApiProductAccessTest.php [hunk 28]
```diff
@@ -306,8 +306,8 @@ protected function developerAppEditFormTest() {
...
+    // Products on on an other user's add/edit form that the other user has
```
Comment (arlina-espinoza) [inline tests/src/Functional/ApiProductAccessTest.php]:
```suggestion
    // Products on another user's add/edit form that the other user has
```

**Review by arlina-espinoza — CHANGES_REQUESTED** [review]
Just a couple of typos/wording changes requested.
Otherwise it looks good! I tested the command using drush and drupal console, and both worked correctly, creating the role with the right permissions:
Screenshot_2019-12-12 arlina-9da89b94-eval - Apigee

File: console/translations/en/apigee_edge.role.create.yml (hunk 27 above)
Comment (cnovak) [reply console/translations/en/apigee_edge.role.create.yml]: Check out my new wording in last commit to see if it helps explain

File: console/translations/en/apigee_edge.role.create.yml [hunk 29]
```diff
@@ -0,0 +1,13 @@
+description: 'Create Apigee role for Drupal'
+help: 'Create a custom Apigee role that limits permissions for Drupal connections to the Apigee API. '
+arguments:
+  org: 'The Apigee Edge org to create the role in'
+  email: 'An Apigee user email address with orgadmin role for this org'
+options:
+  password: 'Password for the Apigee orgadmin user (if not passed in you will be prompted)'
+  base-url: 'Base URL to use, defaults to public cloud URL https://api.enterprise.apigee.com/v1'
+  role-name: 'The role to create in the Apigee Edge org, defaults to "drupalportal"'
+  force: 'Force running of permissions on a role that already exists. Note that permissions are only added, any current permissions not not removed.'
```
Comment (arlina-espinoza) [inline console/translations/en/apigee_edge.role.create.yml:10]:
```suggestion
  force: 'Force running of permissions on a role that already exists. Note that permissions are only added, any current permissions are not removed.'
```

**Review by arlina-espinoza — APPROVED** [review]
@cnovak All LGTM! Just typo suggestion I made.
